crates/proof-of-sql/src/proof_primitive/dory/dynamic_dory_structure.rs

```diff
@@ -125,4 +165,109 @@ mod tests {
+        assert_eq!(width_of_row, 1);
+
+        let width_of_row = full_width_of_row(1);
+        assert_eq!(width_of_row, 2);
+
+        let width_of_row = full_width_of_row(4);
+        assert_eq!(width_of_row, 4);
+
+        let width_of_row = full_width_of_row(7);
+        assert_eq!(width_of_row, 8);
+
+        let width_of_row = full_width_of_row(11);
+        assert_eq!(width_of_row, 8);
+
+        let width_of_row = full_width_of_row(12);
+        assert_eq!(width_of_row, 16);
+
+        let width_of_row = full_width_of_row(13);
+        assert_eq!(width_of_row, 16);
+    }
+    #[test]
+    fn we_can_produce_the_correct_matrix_size() {
+        let (width, height) = matrix_size(2, 0);
+        assert_eq!(width, 2);
+        assert_eq!(height, 2);
+
+        let (width, height) = matrix_size(3, 0);
+        assert_eq!(width, 2);
+        assert_eq!(height, 3);
+
+        let (width, height) = matrix_size(4, 0);
+        assert_eq!(width, 2);
+        assert_eq!(height, 3);
+
+        let (width, height) = matrix_size(16, 0);
+        assert_eq!(width, 4);
+        assert_eq!(height, 6);
+
+        let (width, height) = matrix_size(17, 0);
+        assert_eq!(width, 8);
+        assert_eq!(height, 7);
+
+        let (width, height) = matrix_size(64, 0);
+        assert_eq!(width, 8);
+        assert_eq!(height, 12);
+
+        let (width, height) = matrix_size(71, 0);
+        assert_eq!(width, 16);
+        assert_eq!(height, 13);
+    }
+    #[test]
+    fn we_can_produce_the_correct_matrix_size_with_offset() {
+        let (width, height) = matrix_size(2, 1);
+        assert_eq!(width, 2);
+        assert_eq!(height, 3);
+
+        let (width, height) = matrix_size(16, 1);
+        assert_eq!(width, 8);
+        assert_eq!(height, 7);
+
+        let (width, height) = matrix_size(64, 1);
+        assert_eq!(width, 16);
+        assert_eq!(height, 13);
+
+        let (width, height) = matrix_size(4, 16);
+        assert_eq!(width, 8);
+        assert_eq!(height, 7);
+    }
+    #[test]
+    fn we_can_find_index_from_row_and_column() {
+        for i in 0..1 << 10 {
+            let (row, column) = row_and_column_from_index(i);
+            let index = index_from_row_and_column(row, column);
+            assert_eq!(index, i);
+        }
+    }
+    #[test]
+    fn we_can_find_index_from_row_and_column_not_in_dynamic_dory_structure() {
+        let index = index_from_row_and_column(0, 1);
+        assert_eq!(index, no_index());
+
+        let index = index_from_row_and_column(1, 0);
+        assert_eq!(index, no_index());
+
+        let index = index_from_row_and_column(1, 2);
+        assert_eq!(index, no_index());
+
+        let index = index_from_row_and_column(3, 4);
+        assert_eq!(index, no_index());
+
+        let index = index_from_row_and_column(6, 8);
+        assert_eq!(index, no_index());
+
+        let index = index_from_row_and_column(12, 16);
+        assert_eq!(index, no_index());
+    }
```

Comment:
We could make this more thorough by showing that everything _NOT_ generated by `row_and_column_from_index` returns no index.

Here's Copilot's (AI's) best attempt at that. Obviously needs a lot of work, but my prompt was:
> Can you write a test that will create an IndexSet of all possible row,columns that could be created from the index. Then, test that all the other pairs generate no index.
```suggestion
#[test]
fn test_all_possible_row_column_pairs() {
    use std::collections::HashSet;

    const MAX_INDEX: usize = 1 << 10; // You may need to adjust this value based on your requirements
    let mut valid_pairs = HashSet::new();

    // Collect all valid (row, column) pairs
    for i in 0..MAX_INDEX {
        let (row, column) = row_and_column_from_index(i);
        valid_pairs.insert((row, column));
    }

    // Check that all valid pairs generate a valid index
    for &(row, column) in &valid_pairs {
        let index = index_from_row_and_column(row, column);
        assert_ne!(index, no_index(), "Valid pair ({}, {}) generated no index", row, column);
    }

    // Check that all other pairs generate no_index
    for row in 0..MAX_INDEX {
        for column in 0..full_width_of_row(row) {
            if !valid_pairs.contains(&(row, column)) {
                let index = index_from_row_and_column(row, column);
                assert_eq!(index, no_index(), "Invalid pair ({}, {}) generated a valid index", row, column);
            }
        }
    }
}
```

Reply:
Good idea, added.